compiler/rustc_middle/src/mir/traversal.rs

```diff
@@ -219,6 +215,42 @@ impl<'a, 'tcx> Iterator for Postorder<'a, 'tcx> {
     }
 }
 
+/// Creates an iterator over the `Body`'s basic blocks, that:
+/// - returns basic blocks in a postorder,
+/// - traverses the `BasicBlocks` CFG cache's reverse postorder backwards, and does not cache the
+///   postorder itself.
+pub fn postorder<'a, 'tcx>(body: &'a Body<'tcx>) -> PostorderIter<'a, 'tcx> {
+    let blocks = body.basic_blocks.reverse_postorder();
+    let len = blocks.len();
+    PostorderIter { body, blocks, idx: len }
+}
+
+#[derive(Clone)]
+pub struct PostorderIter<'a, 'tcx> {
+    body: &'a Body<'tcx>,
+    blocks: &'a [BasicBlock],
+    idx: usize,
+}
+
+impl<'a, 'tcx> Iterator for PostorderIter<'a, 'tcx> {
+    type Item = (BasicBlock, &'a BasicBlockData<'tcx>);
+
+    fn next(&mut self) -> Option<(BasicBlock, &'a BasicBlockData<'tcx>)> {
+        if self.idx == 0 {
+            return None;
+        }
+        self.idx -= 1;
+
+        self.blocks.get(self.idx).map(|&bb| (bb, &self.body[bb]))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.idx, Some(self.idx))
+    }
+}
+
+impl<'a, 'tcx> ExactSizeIterator for PostorderIter<'a, 'tcx> {}
```

Comment:
Why do we need to open-code the combinators?

```suggestion
pub fn postorder<'a, 'tcx>(body: &'a Body<'tcx>) -> impl Iterator<Item = (BasicBlock, &'a BasicBlockData<'tcx>)> {
    reverse_postorder(body).rev()
}
```

Reply:
Sure, we don't need it per se, this is the existing RPO one renamed. 

I can remove it, and remove the duplication as you suggest if `reverse_postorder` also becomes a double ended iter.

Reply:
I've pushed a commit to address this.